### ISSUE TITLE
Distillation compute_loss ignores num_items_in_batch and can over-scale gradient accumulation

### ISSUE THREAD

**Issue by Amix29** [MakazhanAlpamys/Soup#720]
## Summary

The custom `_DistillTrainer.compute_loss()` accepts `num_items_in_batch` but
ignores it. It averages CE and KD inside each microbatch and does not explicitly
set `model_accepts_loss_kwargs = False`.

When Transformers classifies the model as accepting loss kwargs and supplies
`num_items_in_batch`, its usual gradient-accumulation compensation can be skipped
even though Soup has not normalized by the global denominator. This can make the
accumulated gradient depend on `gradient_accumulation_steps` rather than only on
the effective batch.

## Reproduction

Verified against current `main` at
`9f1e48b33a5fcda7e621e612ec9305cb52a38e07` with Transformers 5.16.1 and
PyTorch 2.13.0. In an isolated reproduction of the affected Trainer normalization
branch using four equal microbatches:

```text
full-batch gradient norm:       0.2638700074
affected accumulated norm:     1.0554800295
affected / reference ratio:    4.0
explicit /4 control max error: 1.39e-17
```

This demonstrates the normalization contract; it does not imply that every Adam
parameter update is exactly four times larger.

## Impact

For affected models, changing gradient accumulation can silently change gradient
scale and training dynamics despite an equivalent effective batch size. Averaging
each microbatch independently is also not globally token-weighted when sequence
lengths differ.

## Suggested fix

The minimal safe fix for the current mean-reduced loss is to mark the custom
trainer as not consuming loss kwargs:

```python
self.model_accepts_loss_kwargs = False
```

A more complete fix would compute summed CE/KD numerators and normalize by the
global number of supervised labels. That is needed if unequal-length microbatches
must match a single global token mean.

## Acceptance criteria

- [ ] Equivalent effective batches produce matching gradients for GA=1/4/8.
- [ ] Tests cover equal and unequal microbatch lengths.
- [ ] Tests cover response-only labels, padding, an incomplete final accumulation
      window, and distributed denominator handling.
- [ ] The behavior is explicit for every branch of `_DistillTrainer.compute_loss`,
      including sequence KD, MiniLLM, ULD, and ordinary logit KD.

**Comment by MakazhanAlpamys** [comment]
@jagadeepmamidi @dchaudhari7177 — **a collision on this issue, and for the sixth time in five days it is my process that caused it. Neither of you did anything wrong, and I am going to say exactly what happened with timestamps rather than smooth it over.**

## The sequence

- **07:59 UTC** — @jagadeepmamidi asked, on #692's thread, to take **#720**, and asked a scope question before writing any code: minimal `model_accepts_loss_kwargs = False`, or global token normalization as well.
- **I did not answer.**
- **13:18 UTC** — @dchaudhari7177 opened **#742** against this issue.

**#720 had zero comments at that moment.** The claim was made on another issue's thread, so there was nothing here for @dchaudhari7177 to see, and no label either. They opened a pull request against what the repository was showing them: an unclaimed bug.

## What I am changing, because the last fix was not enough

Four days ago I added a `claimed` label after the fifth collision and said a label is the mechanism because comments are invisible in the issue list. That was necessary and it was not sufficient — **a claim made in a different thread is invisible even to me**, which is exactly what happened here. So the rule from now on is: when anyone asks for an issue **anywhere**, I label *that issue* immediately, before answering. Labelling is now the first action, not part of the reply I got to five hours late.

## Where this leaves #720, and @jagadeepmamidi gets the choice

I am not going to hold a complete pull request behind a claim that has no branch behind it — that punishes @dchaudhari7177 for my silence. But I am also not going to decide for you, @jagadeepmamidi, because you did the right thing twice: you asked first, and you asked a scope question instead of guessing.

**If you already have work in progress on this, say so here within 24 hours and I will run the same battery over both** — the way I did for #676/#679 — publish the identical measurements in both threads, and pick on merits rather than on who was first. If you have not started, #742 proceeds on its merits and nothing is lost but your time reading the issue, which I am sorry for.

Either way, **#367 is yours if you want it** and I will label it the moment you say so. It is the closest thing on the board to work you have already proved you do well: `utils.live_eval.load_model_and_tokenizer` takes no quantisation argument, so `soup ship`, `soup diagnose --base-model`, `soup advise --probe-model` and `eval behavior` all load the base in bf16 no matter how the adapter was trained — an NF4 adapter is judged against a base it never saw, and the judge quietly uses double the memory of the training config. That is the same shape as #710: thread a setting through a loader and prove it reaches the call rather than merely being accepted.

## Your scope question, answered — it deserved an answer regardless

**The minimal fix, and only the minimal fix.** `self.model_accepts_loss_kwargs = False` is the whole change, because it hands the compensation back to `Trainer`, which already does it correctly. Global token normalization means re-deriving inside `compute_loss` what the framework computes one level up, and the two would then have to agree forever.

The unequal-microbatch-length wording in the acceptance criteria is about the **test**, not the fix: with equal microbatches the accumulated-gradient ratio is exactly `gradient_accumulation_steps` whether or not the normalization is token-weighted, so an equal-length test cannot distinguish a correct fix from a plausible-looking wrong one. Unequal lengths are what make the assertion discriminating.

@dchaudhari7177 — your PR is under review now; nothing above changes that, and the 24-hour window is a courtesy to a claim I dropped, not a doubt about your work.

**Comment by jagadeepmamidi** [comment]
Thanks for clarifying @MakazhanAlpamys , and no worries. I haven’t started implementation on #720 yet, so I’m happy for #742 to proceed.

I’d like to take #367 instead. The scope looks clear, and I’ll work on threading the quantization setting through the live-eval loader and add tests to verify that the requested numerics actually reach the model-loading path.

**Comment by MakazhanAlpamys** [comment]
Fixed by @dchaudhari7177 in #742, merged as `dd151374`.

`_DistillTrainer` inherited transformers' loss-kwargs path, so with `gradient_accumulation_steps > 1` the accumulated gradient scaled with the step count instead of matching the equivalent full batch. Measured: exactly `4.00000046` at GA=4, i.e. the accumulation factor. Opting out after `super().__init__` restores step-count independence.

Verified before merge on transformers 5.17.0: restoring `model_accepts_loss_kwargs = True` fails both gradient measurements on a **gradient norm ratio**, not on a flag check — the test compiles the real `_DistillTrainer` class body out of `distill.py` and runs `Trainer.training_step` on a real model. A behaviourally identical respelling (`= not True`) correctly still passes, so the check is on behaviour rather than on source text.

**One residual this does not close, stated so it is not lost:** a per-microbatch mean weights microbatches equally rather than tokens, so with unequal sequence lengths the fix restores step-count independence but not the full-batch number. That is a separate change and I am filing it.

Try:

```
soup train --config your.yaml   # task: distill, with gradient_accumulation_steps > 1
```
